File: app/src/main/java/com/silentgames/silent_planet/logic/TurnHandler.java
package com.silentgames.silent_planet.logic;

import android.app.Activity;
import android.content.Context;
import android.widget.Toast;

import com.silentgames.silent_planet.model.fractions.Fractions;
import com.silentgames.silent_planet.model.fractions.FractionsEnum;

/**
 * Created by gidroshvandel on 26.09.16.
 */
public class TurnHandler {

    private static int turnCount;
    private static FractionsEnum fraction;

    public TurnHandler() {
    }

    public static void turnCount(){
        int turnCount = getTurnCount();
        setTurnCount(turnCount ++);
        nextPlayer();
    }

    public static void start(Fractions fraction){

        setFraction(fraction.getFractionsEnum());
    }

    public static void setPlayable(Fractions fraction){
        fraction.setPlayable(true);
    }

    private static void nextPlayer(){
        setFraction(getFraction().next());
    }

    public static int getTurnCount() {
        return turnCount;
    }

    public static void setTurnCount(int turnCount) {
        TurnHandler.turnCount = turnCount;
    }

    public static FractionsEnum getFraction() {
        return fraction;
    }

    public static void setFraction(FractionsEnum fraction) {
        TurnHandler.fraction = fraction;
    }
}

File: app/src/main/java/com/silentgames/silent_planet/logic/TurnHandler.kt
package com.silentgames.silent_planet.logic

import com.silentgames.silent_planet.model.fractions.FractionsType
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.channels.BroadcastChannel
import kotlinx.coroutines.channels.sendBlocking
import kotlinx.coroutines.flow.asFlow

/**
 * Created by gidroshvandel on 26.09.16.
 */
object TurnHandler {

    @ExperimentalCoroutinesApi
    private val channel = BroadcastChannel<FractionsType>(1)

    private var turnCount: Int = 0
    lateinit var fractionType: FractionsType

    @ExperimentalCoroutinesApi
    fun turnCount() {
        turnCount++
        nextPlayer()
        channel.sendBlocking(fractionType)
    }

    @ExperimentalCoroutinesApi
    @FlowPreview
    fun getFlow() = channel.asFlow()

    fun start(fraction: FractionsType) {
        this.fractionType = fraction
    }

    private fun nextPlayer() {
        fractionType = fractionType.next()
    }
}
